File: module_3/src/main/java/alevel/service/UserService.java
package alevel.service;

import alevel.dto.table.ResponseAbstTablePage;
import alevel.dto.table.impl.ResponseUserTablePage;
import alevel.model.impl.Account;
import alevel.model.impl.Operation;
import alevel.model.impl.User;
import org.apache.commons.csv.CSVPrinter;

import javax.servlet.http.HttpServletResponse;
import java.util.List;

public interface UserService {

    List<User> findAll();
    void create(User t);
    void update(User t);
    void delete(Long id);
    User findById(Long id);

    public void createAccount(Account account, Long userId);

    public void updateAccount(Account account);

    public void deleteAccount(Long accountId);

    public Account findAccountById(Long id);

    public Account findAccountByNameAccount(String nameAccount);

    public List<Account> findAllAccountsByUserId(Long id);

    public void createOperation(Operation operation);

    public void updateOperation(Operation operation);

    public void deleteOperation(Long id);

    public Operation findOperationById(Long id);

    public Operation findByAccountId(Long accountId);

    public List<Operation> findAllByAccountId(Long id);

     boolean sendMoneyToUser(Long senderId, Long recipientId, Long acccountSenderId, Long accountRecipientId, long summa) throws Exception;

     ResponseUserTablePage findAllWithSortColumn(int page, int showEntity, String column, String sort);

     ResponseAbstTablePage findAllAccountForUserListPage(int page, int showEntity, String columnSort, String sort, Long userId);

    void exportAccountOperationByUserToCSV(HttpServletResponse response ,Long userId) throws Exception;
}

File: module_3/frontBank/src/app/service/user.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpErrorResponse} from "@angular/common/http";
import {environment} from "../../environments/environment";
import {Observable} from "rxjs";
import {User} from "../model/user";
import {Account} from "../model/account";
import {Operation} from "../model/operation";
import {ResponseUserTablePage} from "../dto/response-user-table-page";
import {ResponseTablePage} from "../dto/response-table-page";

@Injectable({
  providedIn: 'root'
})
export class UserService {

 private host = environment.host;

  constructor(private http: HttpClient) {
  }

  public findAll(): Observable<User[] | HttpErrorResponse> {
    return this.http.get<User[] | HttpErrorResponse>(`${this.host}/user/list-user`);
  }


  public create(user: User): Observable<void> {
    return this.http.post<void>(`${this.host}/user/create_user`, user);
  }


  public update(user: User): Observable<void> {
    return this.http.put<void>(`${this.host}/user/update_user`, user);
  }

  public delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.host}/user/delete_user/${id}`);
  }

  public findUserById(id: number): Observable<User | HttpErrorResponse> {
    return this.http.get<User | HttpErrorResponse>(`${this.host}/user/find_user/${id}`);
  }

  public createAccount(account: Account, userId: number): Observable<void> {
    return this.http.post<void>(`${this.host}/user/create_account/${userId}`, account);
  }

  public updateAccount(account: Account): Observable<void> {
    return this.http.put<void>(`${this.host}/user/update_account`, account);
  }


  public deleteAccount(accountId: number): Observable<void> {
    return this.http.delete<void>(`${this.host}/user/delete_account/${accountId}`);
  }

  public findAccountById(id: number): Observable<Account | HttpErrorResponse> {
    return this.http.get<Account | HttpErrorResponse>(`${this.host}/user/find_account/${id}`);
  }

  public findAccountByNameAccount(nameAccount: string): Observable<Account | HttpErrorResponse> {
    return this.http.get<Account | HttpErrorResponse>(`${this.host}/user/find_account_by_name/${nameAccount}`);
  }

  public findAllAccountsByUserId(userId: number): Observable<Account[] | HttpErrorResponse> {
    return this.http.get<Account[] | HttpErrorResponse>(`${this.host}/user/list_accounts_user/${userId}`);
  }

  public findOperationById(id: number): Observable<Operation | HttpErrorResponse> {
    return this.http.get<Operation | HttpErrorResponse>(`${this.host}/user/find_operation/${id}`);
  }

  public findByAccountId(accountId: number): Observable<Operation | HttpErrorResponse> {
    return this.http.get<Operation | HttpErrorResponse>(`${this.host}/user/find_operation_account/${accountId}`);
  }


  public sendMoney(senderId: number,
                   recipientId: number,
                   acccountSenderId: number,
                   accountRecipientId: number,
                   summa: number): Observable<boolean | HttpErrorResponse> {
    return this.http.get<boolean | HttpErrorResponse>(`${this.host}/user/send/${senderId}/${recipientId}/${acccountSenderId}/${accountRecipientId}/${summa}`);
  }

  public findAllUserListPage(page: number, showEntity: number, column: string, sort: string): Observable<ResponseTablePage<User> | HttpErrorResponse> {
    return this.http.get<ResponseTablePage<User> | HttpErrorResponse>(`${this.host}/user/limit-list/users/${page}/${showEntity}/${column}/${sort}`);
  }

  public findAllAccountForUserListPage(page: number, showEntity: number, column: string, sort: string, userId: number): Observable<ResponseTablePage<Account> | HttpErrorResponse> {
    return this.http.get<ResponseTablePage<Account> | HttpErrorResponse>(`${this.host}/user/limit-accounts_user/users/${page}/${showEntity}/${column}/${sort}/${userId}`);
  }

  public saveSelectUserTolocalCache(user: User) {
    localStorage.setItem('saveUser', JSON.stringify(user));
  }

  public getsaveUserFromLocalCache() {
    return JSON.parse(localStorage.getItem('saveUser'));
  }

  public downloadOperationCSV(userId: number): Observable<any>{
  return this.http.get(`${this.host}/user/download/csv/${userId}`, { responseType: 'blob'});
}
}
